File: 935.knight-dialer.py
#
# @lc app=leetcode id=935 lang=python3
#
# [935] Knight Dialer
#

# @lc code=start
class Solution:
    '''
    This look like a matrix multiplication problem
    At least, we see that there are well defined transitions possible
    1 -> 6, 8
    2 -> 7, 9
    3 -> 4, 8
    4 -> 3, 9, 0
    5 -> {}
    6 -> 1, 7, 0
    7 -> 2, 6
    8 -> 1, 3
    9 -> 2, 4
    0 -> 4, 6
    The transition matrix M is well defined
    Then we can break down n into power of 2s. (n >> i = 1 or not)
    And just sum up the matrix contribution each time it is squared    
    '''

    def matrix_multiply(self, A, C):
        n = len(A)
        B = [[0 for i in range(n)] for j in range(n)]
        for i in range(n):
            for j in range(n):
                for k in range(n):
                    B[i][j] += A[i][k]*C[k][j] 
                    B[i][j] %= self.MOD
        return B
    
    def knightDialer(self, n: int) -> int:
        self.MOD = pow(10, 9) + 7
        M = [[0,0,0,0,1,0,1,0,0,0],
             [0,0,0,0,0,0,1,0,1,0],
             [0,0,0,0,0,0,0,1,0,1],
             [0,0,0,0,1,0,0,0,1,0],
             [1,0,0,1,0,0,0,0,0,1],
             [0,0,0,0,0,0,0,0,0,0],
             [1,1,0,0,0,0,0,1,0,0],
             [0,0,1,0,0,0,1,0,0,0],
             [0,1,0,1,0,0,0,0,0,0],
             [0,0,1,0,1,0,0,0,0,0]]
        seed = [[1 if i == j else 0 for i in range(10)] for j in range(10)]      
        
        for i in range((n-1).bit_length()):
            if (n - 1) & (1 << i) != 0:
                seed = self.matrix_multiply(seed, M)
            M = self.matrix_multiply(M, M)
        
        result = 0    
        for j in range(10):
            for k in range(10):
                result += seed[j][k]
                result %= self.MOD
        return result
# @lc code=end

File: 935.knight-dialer.cpp
/*
 * @lc app=leetcode id=935 lang=cpp
 *
 * [935] Knight Dialer
 */

// @lc code=start
#include <cmath>
#include <vector>

class Solution {
private:
    long long MOD =  (long long)1000000007;
public:

    std::vector<std::vector<long long>> MatrixMultiply(const std::vector<std::vector<long long>>& A, const std::vector<std::vector<long long>>& C) {
        int n = A.size();
        std::vector<std::vector<long long>> B(n, std::vector<long long>(n, (long long)0));

        for (int i = 0; i < n; i++) {
            for (int j = 0; j < n; j++) {
                for (int k = 0; k < n; k++) {
                    B[i][j] += A[i][k] * C[k][j];
                    B[i][j] %= MOD;
                }
            }
        }

        return B;
    }

    int bitLength(int bits){
        int size = 0;
        while(bits != 0){
            bits >>= 1;
            size++;
        }
        return size;       
    }

    int knightDialer(int n) {

        std::vector<std::vector<long long>> M = {
            {0, 0, 0, 0, 1, 0, 1, 0, 0, 0},
            {0, 0, 0, 0, 0, 0, 1, 0, 1, 0},
            {0, 0, 0, 0, 0, 0, 0, 1, 0, 1},
            {0, 0, 0, 0, 1, 0, 0, 0, 1, 0},
            {1, 0, 0, 1, 0, 0, 0, 0, 0, 1},
            {0, 0, 0, 0, 0, 0, 0, 0, 0, 0},
            {1, 1, 0, 0, 0, 0, 0, 1, 0, 0},
            {0, 0, 1, 0, 0, 0, 1, 0, 0, 0},
            {0, 1, 0, 1, 0, 0, 0, 0, 0, 0},
            {0, 0, 1, 0, 1, 0, 0, 0, 0, 0}
        };

        std::vector<std::vector<long long>> seed(10, std::vector<long long>(10, (long long)0));
        for (int i = 0; i < 10; i++) {
            seed[i][i] = 1;
        }

        for (int i = 0; i < bitLength(n - 1); i++) {
            if ((n - 1) & (1 << i)) {
                seed = MatrixMultiply(seed, M);
            }
            M = MatrixMultiply(M, M);
        }

        long long result = 0;
        for (int j = 0; j < 10; j++) {
            for (int k = 0; k < 10; k++) {
                result += seed[j][k];
                result %= MOD;
            }
        }

        return (int)result;
    }
};

// @lc code=end

